lib/flame/code_sync.ex

```diff
@@ -0,0 +1,311 @@
+        if !Enum.empty?(code.apps_to_start),
+          do: log_verbose("sending apps_to_start: #{inspect(code.apps_to_start)}")
+
+        [:verbose]
+      else
+        []
+      end
+
+    out_stream =
+      if code.changed_paths != [] do
+        out_path = Path.join([code.tmp_dir.(), "flame_parent_code_sync_#{code.id}.tar.gz"])
+        dirs = for path <- code.changed_paths, uniq: true, do: String.to_charlist(path)
+        {:ok, tar} = :erl_tar.open(out_path, [:write, :compressed])
+        for dir <- dirs, do: :erl_tar.add(tar, dir, trim_leading_slash(dir), verbose)
+        :ok = :erl_tar.close(tar)
+
+        File.stream!(out_path, [], 2048)
+      end
+
+    %PackagedStream{
+      id: code.id,
+      tmp_dir: code.tmp_dir,
+      extract_dir: code.extract_dir,
+      sync_beam_hashes: code.sync_beam_hashes,
+      changed_paths: code.changed_paths,
+      deleted_paths: code.deleted_paths,
+      purge_modules: code.purge_modules,
+      apps_to_start: code.apps_to_start,
+      stream: out_stream,
+      verbose: code.verbose
+    }
+  end
+
+  defp trim_leading_slash([?/ | path]), do: path
+  defp trim_leading_slash([_ | _] = path), do: path
+
+  def extract_packaged_stream(%PackagedStream{} = pkg) do
+    if pkg.stream do
+      verbose = if pkg.verbose, do: [:verbose], else: []
+      extract_dir = pkg.extract_dir.()
+      target_tmp_path = Path.join([pkg.tmp_dir.(), "flame_child_code_sync_#{pkg.id}.tar.gz"])
+      flame_stream = File.stream!(target_tmp_path)
+      # transfer the file
+      Enum.into(pkg.stream, flame_stream)
+
+      # extract tar
+      :ok = :erl_tar.extract(target_tmp_path, [{:cwd, extract_dir}, :compressed | verbose])
+
+      # add code paths
+      :ok = add_code_paths_from_tar(pkg, extract_dir)
+
+      File.rm!(target_tmp_path)
+
+      # purge any deleted modules
+      for mod <- pkg.purge_modules do
+        if pkg.verbose && !Enum.empty?(pkg.purge_modules),
+          do: log_verbose("purging #{inspect(pkg.purge_modules)}")
+
+        :code.purge(mod)
+        :code.delete(mod)
+      end
+
+      # delete any deleted code paths, and prune empty dirs
+      for del_path <- pkg.deleted_paths do
+        File.rm!(del_path)
+        ebin_dir = Path.dirname(del_path)
+
+        if File.ls!(ebin_dir) == [] do
+          if pkg.verbose, do: log_verbose("deleting path #{ebin_dir}")
+          File.rm_rf!(ebin_dir)
+          :code.del_path(String.to_charlist(ebin_dir))
+        end
+      end
+
+      # reload any changed code
+      reloaded = :c.lm()
+      if pkg.verbose && !Enum.empty?(reloaded), do: log_verbose("reloaded #{inspect(reloaded)}")
+    end
+
+    # start any synced apps
+    if !Enum.empty?(pkg.apps_to_start) do
+      {:ok, started} = Application.ensure_all_started(pkg.apps_to_start)
+      if pkg.verbose, do: log_verbose("started #{inspect(started)}")
+    end
+
+    :ok
+  end
+
+  def rm_packaged_stream!(%PackagedStream{} = pkg) do
+    if pkg.stream, do: File.rm!(pkg.stream.path)
```

Comment:
If we are using tmp, we probably don't care if we can't remove the file.

```suggestion
  def rm_packaged_stream(%PackagedStream{} = pkg) do
    if pkg.stream, do: File.rm(pkg.stream.path)
```